### ISSUE TITLE
Visual Studio Unreachable Code Analyzer incorrectly alerts with `catch when` is in a `for` loop

### ISSUE THREAD

**Issue by mgaffigan** [dotnet/roslyn-analyzers#3282]
#### Analyzer package

Not sure.  Given warning CS0162, I presume it to be an analyzer inside the compiler.

#### Package Version

Microsoft Visual Studio Enterprise 2019
Version 16.4.3
VisualStudio.16.Release/16.4.3+29709.97
Microsoft .NET Framework
Version 4.8.03752

Installed Version: Enterprise

C# Tools   3.4.1-beta4-19610-02+c4e5d138903b899477649a17f197abd2bcb22f9e

#### Diagnostic ID

[CS0162: Unreachable Code Detected](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0162)

#### Repro steps

1. Place try/catch with `when` inside for loop

```
    static void Main(string[] args)
    {
        for (int i = 0; i < 10; i++)
        //                      ^^^ Warning CS0162 Unreachable code detected
        {
            try
            {
                break;
            }
            catch when (i < 9)
            {
                throw;
            }
        }
        Console.WriteLine("Hello");
    }
```

2. Intellisense underline occurs for `i++` in `for` with CS0162 Unreachable Code

#### Expected behavior

No warning is given

#### Actual behavior

Warning is given.

**Comment by mavasani** [comment]
@mgaffigan This issue needs to be ported to https://github.com/dotnet/roslyn/issues/new. This is a compiler warning, not an analyzer warning.
